catalog/ui/catalog-ui-search/src/main/java/org/codice/ddf/catalog/ui/query/suggestion/MgrsCoordinateProcessor.java

```diff
@@ -0,0 +1,340 @@
+
+  // This key tells the UI that the geo is on the suggestion itself
+  private static final String LITERAL_SUGGESTION_ID = "LITERAL-MGRS";
+
+  private static final String ZERO = "0";
+
+  private static final String WHITESPACE = " ";
+
+  /**
+   * An {@link MgrsPartType} identifies one of the three different component strings that are glued
+   * together to form a valid MGRS coordinate string, including a category {@link #NONE} to be used
+   * when the string is unsuitable for any part of an MGRS coordinate or the type has not been
+   * provided yet.
+   *
+   * <p>Each type corresponds to a regex defined above.
+   */
+  private enum MgrsPartType {
+    NONE,
+    ZONE,
+    HUNDRED_KM,
+    NUMERIC
+  }
+
+  private final CoordinateSystemTranslator translator = new CoordinateSystemTranslatorImpl();
+
+  /**
+   * Given a list of {@link Suggestion}s and the query that yielded them, enhance the list with
+   * additional suggestions based upon the presence of coordinate literals in the query string.
+   *
+   * @param results the list of results to enhance.
+   * @param query the query string to search for coordinate literals.
+   */
+  public void enhanceResults(List<Suggestion> results, String query) {
+    final LiteralSuggestion literal = getMgrsSuggestion(query);
+    if (literal != null && literal.hasGeo()) {
+      LOGGER.trace("Adding the MGRS suggestion to results");
+      results.add(0, literal);
+    }
+  }
+
+  /**
+   * Build the MGRS suggestion.
+   *
+   * @param query the query string to search for coordinate literals.
+   * @return a suggestion object with geometry attached to it; a {@link BoundingBox} if only one
+   *     coordinate was detected, or a collection of {@link LatLon} center points otherwise. If no
+   *     MGRS was detected in the query, {@code null} is returned.
+   */
+  @Nullable
+  private LiteralSuggestion getMgrsSuggestion(String query) {
+    final List<String> matches = getMgrsCoordinateStrings(query);
+    if (matches.isEmpty()) {
+      LOGGER.debug("No valid MGRS strings were found or could be inferred from query [{}]", query);
+      return null;
+    }
+
+    final List<UsngCoordinate> mgrsCoordinates = new ArrayList<>();
+    final StringBuilder nameBuilder = new StringBuilder("MGRS: ");
+    for (String match : matches) {
+      UsngCoordinate coordinate = parseMgrsString(match);
+      if (coordinate != null) {
+        mgrsCoordinates.add(coordinate);
+        nameBuilder.append("[ ").append(coordinate.toMgrsString()).append(" ] ");
+      }
+    }
+
+    if (mgrsCoordinates.size() == 1) {
+      final UsngCoordinate mgrsCoord = mgrsCoordinates.get(0);
+      final BoundingBox bb = translator.toBoundingBox(mgrsCoord);
+      return new LiteralSuggestion(
+          LITERAL_SUGGESTION_ID, nameBuilder.toString(), latLonFromBoundingBox(bb));
+    }
+    return new LiteralSuggestion(
+        LITERAL_SUGGESTION_ID,
+        nameBuilder.toString(),
+        mgrsCoordinates
```

Comment:
✏️ Consider extracting the result of the stream to a local variable to reduce nesting.